Sources/_StringProcessing/Engine/MEQuantify.swift

```diff
@@ -1,166 +1,512 @@
-    var maxExtraTrips = payload.maxExtraTrips
+extension String {
+  /// Run the quant loop, using the supplied matching closure
+  ///
+  /// NOTE: inline-always to help elimiate the closure overhead,
+  /// simplify some of the looping structure, etc.
+  @inline(__always)
+  fileprivate func _runQuantLoop(
+    at currentPosition: Index,
+    limitedBy end: Index,
+    minMatches: UInt64,
+    maxMatches: UInt64,
+    produceSavePointRange: Bool,
+    isScalarSemantics: Bool,
+    _ doMatch: (
+     _ currentPosition: Index, _ limitedBy: Index, _ isScalarSemantics: Bool
+    ) -> Index?
+  ) -> (next: Index, savePointRange: Range<Index>?)? {
+    var currentPosition = currentPosition
+
+    // The range of backtracking positions to try. For zero-or-more, starts
+    // before any match happens. Always ends before the final match, since
+    // the final match is what is tried without backtracking. An empty range
+    // is valid and means a single backtracking position at rangeStart.
+    var rangeStart = currentPosition
+    var rangeEnd = currentPosition
+
+    var numMatches = 0
 
-    while trips < payload.minTrips {
-      guard let next = _doQuantifyMatch(payload) else {
-        signalFailure()
-        return false
+    while numMatches < maxMatches {
+      guard let next = doMatch(
+        currentPosition, end, isScalarSemantics
+      ) else {
+        break
+      }
+      numMatches &+= 1
+      if numMatches == minMatches {
+        // For this loop iteration, rangeEnd will actually trail rangeStart by
+        // a single match position. Next iteration, they will be equal
+        // (empty range denoting a single backtracking point). Note that we
+        // only ever return a range if we have exceeded `minMatches`; if we
+        // exactly match `minMatches` there is no backtracking positions to
+        // remember.
+        rangeStart = next
       }
+      rangeEnd = currentPosition
       currentPosition = next
-      trips += 1
+      assert(currentPosition > rangeEnd)
     }
 
-    if maxExtraTrips == 0 {
-      // We're done
-      return true
+    guard numMatches >= minMatches else {
+      return nil
     }
 
-    guard let next = _doQuantifyMatch(payload) else {
-      return true
+    guard produceSavePointRange && numMatches > minMatches else {
+      // No backtracking positions to try
+      return (currentPosition, nil)
     }
-    maxExtraTrips = maxExtraTrips.map { $0 - 1 }
+    assert(rangeStart <= rangeEnd)
 
-    // Remember the range of valid positions in case we can create a quantified
-    // save point
-    let rangeStart = currentPosition
-    var rangeEnd = currentPosition
-    currentPosition = next
+    // NOTE: We can't assert that rangeEnd trails currentPosition by exactly
+    // one position, because newline-sequence in scalar semantic mode still
+    // matches two scalars
 
-    while true {
-      if maxExtraTrips == 0 { break }
+    return (currentPosition, rangeStart..<rangeEnd)
```

Comment:
With the assertion above, should this be `Range(uncheckedBounds:)` instead of `..<`?

Reply:
Yes, that would work.